File: projects/06/go-assembler/instruction.go
package main

import (
	"strings"
)

type Instruction struct {
	instructionType InstructionType
	destination     string
	computation     string
	jump            string
	rawValue        string
}

func NewEmptyInstruction() Instruction {
	return Instruction{}
}

func (i *Instruction) ProcessInstruction(rawValue string) {
	i.rawValue = rawValue

	if i.rawValue == "" || strings.HasPrefix(i.rawValue, "/") {
		return
	} else if strings.HasPrefix(i.rawValue, "@") {
		i.instructionType = InstructionA
	} else if strings.HasPrefix(i.rawValue, "(") {
		i.instructionType = InstructionL
	} else {
		i.instructionType = InstructionC
		i.rawValue = strings.Split(i.rawValue, "//")[0]
		i.getSymbolicCInstruction()
	}
}

func (i *Instruction) getSymbolicCInstruction() {
	if strings.Contains(i.rawValue, ";") {
		parts := strings.Split(i.rawValue, ";")
		i.computation = strings.TrimSpace(parts[0])
		i.jump = strings.TrimSpace(parts[1])
		i.destination = "null"
	} else {
		parts := strings.Split(i.rawValue, "=")
		i.computation = strings.TrimSpace(parts[1])
		i.destination = strings.TrimSpace(parts[0])
		i.jump = "null"
	}
}

func (i *Instruction) GetSymbol() string {
	if strings.HasPrefix(i.rawValue, "(") {
		endIndex := strings.Index(i.rawValue, ")")
		return i.rawValue[1:endIndex]
	}
	return i.rawValue[1:]
}

func (i *Instruction) GetDestination() string {
	return i.destination
}

func (i *Instruction) GetComputation() string {
	return i.computation
}

func (i *Instruction) GetJump() string {
	return i.jump
}

func (i *Instruction) GetInstructionType() InstructionType {
	return i.instructionType
}

File: projects/06/assembler/instruction.py
from utils import InstructionType


class Instruction:

    def __init__(self) -> None:
        self.instruction_type: InstructionType = ''
        self.destination: str = 'null'
        self.computation: str = '0'
        self.jump: str = 'null'
        self.raw_value: str = ''

    def process_instruction(self, raw_value: str) -> None:
        self.raw_value = raw_value
        
        if not self.raw_value or self.raw_value.startswith('/'):
            return

        elif self.raw_value.startswith('@'):
            self.instruction_type = InstructionType.A
      
        elif self.raw_value.startswith('('):
            self.instruction_type = InstructionType.L
        
        else:
            self.instruction_type = InstructionType.C
            self.raw_value = self.raw_value.split("//")[0]
            self._get_symbolic_c_instruction()
    
    def _get_symbolic_c_instruction(self) -> None:

        if ';' in self.raw_value:
            parts = self.raw_value.split(';')
            self.computation = parts[0].strip()
            self.jump = parts[1].strip()
            self.destination = 'null'
        else:
            parts = self.raw_value.split('=')
            self.computation = parts[1].strip()
            self.destination = parts[0].strip()
            self.jump = 'null'
        
    def get_symbol(self) -> str:
        if self.raw_value.startswith('('):
            end_index = self.raw_value.find(')')
            return self.raw_value[1:end_index]
     
        return self.raw_value[1:]

    def get_destination(self) -> str:
        return self.destination

    def get_computation(self) -> str:
        return self.computation
    
    def get_jump(self) -> str:
        return self.jump

    def get_instruction_type(self) -> InstructionType:
        return self.instruction_type
